### PR TITLE
mgr/cephadm: allow setting prometheus retention time

### PR DESCRIPTION
**Pull request by adk3798**
When we deploy Prometheus server, we don't provide any
ability to define the tsdb retention time - so it defaults to 15d.

This change adds a field that can be passed in a prometheus service
spec that will be passed as an arg to the --storage.tsdb.retention.time
parameter for the prometheus daemon.

Fixes: https://tracker.ceph.com/issues/54308

Signed-off-by: Adam King <adking@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/cephadm/services/monitoring.rst [hunk 1]
```diff
@@ -328,6 +328,32 @@ Due to performance reasons, monitoring of RBD images is disabled by default. For
 :ref:`prometheus-rbd-io-statistics`. If disabled, the overview and details dashboards will stay empty in Grafana
 and the metrics will not be visible in Prometheus.
 
+Setting up Prometheus
+-----------------------
+
+Setting Prometheus Retention Time
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Cephadm provides the option to set the prometheus tdsb retention time using
```
Comment (anthonyeleven) [inline doc/cephadm/services/monitoring.rst:337]: Prometheus TSDB

File: doc/cephadm/services/monitoring.rst [hunk 2]
```diff
@@ -328,6 +328,32 @@ Due to performance reasons, monitoring of RBD images is disabled by default. For
...
+
+    service_type: prometheus
+    placement:
+      count: 1
+    spec:
+      retention_time: "1y"
+
+.. note::
+
+  If you already had prometheus daemon(s) deployed before and are updating an
```
Comment (anthonyeleven) [inline doc/cephadm/services/monitoring.rst:352]: Prometheus capitalization here and below (thrice)

**Review by anthonyeleven — CHANGES_REQUESTED** [review]
Nice!  Recent Prometheus releases also have the ability to limit retention with `--storage.tsdb.retention.size` (cf. https://prometheus.io/docs/prometheus/latest/storage/)

Might it be possible to add that as well?  Should be straightforward to clone your work here.

File: doc/cephadm/services/monitoring.rst [hunk 3]
```diff
@@ -328,6 +328,32 @@ Due to performance reasons, monitoring of RBD images is disabled by default. For
...
+a ``retention_time`` field in the prometheus service spec. The value defaults
+to 15 days (15d). If you would like a different value, such as 1 year (1y) you
+can apply a service spec similar to:
+
+.. code-block:: yaml
+
+    service_type: prometheus
+    placement:
+      count: 1
+    spec:
```
Comment (rkachach) [inline doc/cephadm/services/monitoring.rst:347]: Please, just ignore if it doesn't apply but I remember the configuration section was called something like _config_

**Review by rkachach — APPROVED** [review]
LGTM

File: doc/cephadm/services/monitoring.rst (hunk 3 above)
Comment (adk3798) [reply doc/cephadm/services/monitoring.rst:347]: config is for telling cephadm to set config options, spec is for specifying fields that aren't part of the generic service spec but are included in a specific service's spec. Eg. `rgw_frontend_port` for rgw would go in the spec section.

**Comment by adk3798** [comment]
> Nice! Recent Prometheus releases also have the ability to limit retention with `--storage.tsdb.retention.size` (cf. https://prometheus.io/docs/prometheus/latest/storage/)
> 
> Might it be possible to add that as well? Should be straightforward to clone your work here.

probably, I'll just need to verify what versions of prometheus have this to make sure our default version is new enough. I'll do it in a follow up PR if it's viable.

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
jenkins test api

**Comment by adk3798** [comment]
jenkins test dashboard cephadm

**Comment by adk3798** [comment]
@anthonyeleven mind giving this another look

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-08-30_17:06:47-orch:cephadm-wip-adk-testing-2022-08-29-1644-distro-default-smithi/
and reruns https://pulpito.ceph.com/adking-2022-09-01_12:53:01-orch:cephadm-wip-adk-testing-2022-08-29-1644-distro-default-smithi/

Failures tracked by:

- https://tracker.ceph.com/issues/57287
- https://tracker.ceph.com/issues/57374
- https://tracker.ceph.com/issues/57371
- https://tracker.ceph.com/issues/56573
- https://tracker.ceph.com/issues/55347

the only one that is new and surprising on this run were the iscsi test failures (https://tracker.ceph.com/issues/57371). As a result, I'm not going to merge any PRs in this run that seem related to iscsi. Other PRs I think will still be okay to merge though.

**Comment by adk3798** [comment]
NOTE: last force push only fixed a merge conflict in a unit test. Shouldn't need another teuth run for this

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
jenkins test dashboard cephadm

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-09-02_11:54:06-orch:cephadm-wip-adk-testing-2022-09-01-2029-distro-default-smithi/

Failures tracked by

- https://tracker.ceph.com/issues/57371
- https://tracker.ceph.com/issues/55986
- https://tracker.ceph.com/issues/57287
- https://tracker.ceph.com/issues/57374

All failures are also present on main branch and not caused by PRs in the run

**Review by anthonyeleven — APPROVED** [review]

**Comment by anthonyeleven** [comment]
Ack, please do as I personally find that the more useful option.

**Comment by adk3798** [comment]
> Ack, please do as I personally find that the more useful option.

created https://tracker.ceph.com/issues/57422 for tracking that

**Comment by anthonyeleven** [comment]
> > Ack, please do as I personally find that the more useful option.
> 
> created https://tracker.ceph.com/issues/57422 for tracking that

❤️

**Pull request merged**